### PR TITLE
[v0.11 feature] Per-chain Relays To Token Multiplier

### PR DESCRIPTION
**Pull request by msmania**
## Description
This patch introduces a new network parameter `pos/RelaysToTokensMultiplierMap` so that we can set a custom RTTM per chain.  The existing parameter `pos/RelaysToTokensMultiplier` serves the default RTTM if a custom RTTM is not set.

This change is consensus-breaking.  The new behavior is put behind a new feature key `RTTM2`.  The new parameter is kept unavailable until activation.

The patch also contains two tests:
- `Test_RTTM2_ChangeParamValue` to make sure the new parameter is kept unavailable until activating the `RTTM2` feature key.
- `TestKeeper_RewardForRelaysPerChain` to make sure the amount of relay rewards are calculated with a custom RTTM in the new parameter if it's set.

### Summary generated by Reviewpad on 07 Sep 23 00:52 UTC
This pull request contains the following changes:

1. The addition of a new test function called `TestKeeper_RewardForRelaysPerChain`. This test function verifies the rewards generated for different scenarios related to relays per chain.

2. Modifications to the file "expectedKeepers.go" were made. A new function called "RewardForRelaysPerChain" was added, which calculates and returns a value of type `sdk.BigInt`.

3. Changes were made to the file "module.go" related to activating additional parameters and setting a minimum signed per window value.

4. The file "keeper_test.go" includes the addition of a new test function called "Test_RTTM2_ChangeParamValue", which tests the change of a parameter value after a certain activation height.

5. A new file "proof.go" was modified to enhance the functionality and reliability of the `ExecuteProof` function in handling relay proofs.

6. The file "param.go" now includes additional tracking of parameters introduced after genesis and skips writing them to the database until the corresponding feature activation height is reached.

7. The file "nodes.go" includes changes to the `AwardCoinsForRelays` function, enabling it to award coins for relays completed per chain.

8. Modifications were made to the `PosKeeper` interface in the file "expectedKeepers.go", adding a new function called `RewardForRelaysPerChain`.

9. The file "service_test.go" includes changes to the `RewardForRelays` method in the `MockPosKeeper` struct, allowing coins to be awarded for relays completed per chain.

10. The file "abci.go" in the `x/nodes/keeper` package includes import statement additions and changes to variable names.

11. The go.sum file shows updates to dependencies related to the "github.com/pokt-network/tendermint" package.

These changes enhance functionality, introduce new features, and improve reliability in various parts of the codebase.

### CODE REVIEW

**Comment by POKT-Discourse** [comment]
This pull request has been mentioned on **Pocket Network Forum**. There might be relevant details there:

https://forum.pokt.network/t/unleashing-the-potential-of-pocket/4720/1

**Comment by POKT-Discourse** [comment]
This pull request has been mentioned on **Pocket Network Forum**. There might be relevant details there:

https://forum.pokt.network/t/unleashing-the-potential-of-pocket-as-universal-rpc-provider/4797/1

File: app/tx_test.go [hunk 1]
```diff
@@ -1231,6 +1232,144 @@ func TestChangeParamspip22afterActivationHeight(t *testing.T) {
 	}
 }
 
+func Test_RTTM2_ChangeParamValue(t *testing.T) {
+	cdc := memCodec()
+	ParamName := "pos/RelaysToTokensMultiplierMap"
```
Comment (Olshansk) [inline app/tx_test.go]: NIT: Why are these capitalized? They're not constants and not exposed publically.

I even asked ChatGPT

Screenshot 2023-11-28 at 6 40 38 PM

File: app/tx_test.go [hunk 2]
```diff
@@ -1231,6 +1232,144 @@ func TestChangeParamspip22afterActivationHeight(t *testing.T) {
...
+		"0001": 12345,
+	}
+	ExpectedValue, err := cdc.MarshalJSON(ParamValue)
+	assert.Nil(t, err)
+
+	testModeTemp := codec.TestMode
+	upgradeHeightTemp := codec.UpgradeHeight
+
+	// Prepare governance parameters
+	// Includes codec upgrade, validator split, non-custodial upgrade,
```
Comment (Olshansk) [inline app/tx_test.go]: `validator split, non-custodial upgrade, and block size reduction`

Is this a copy-pasta error or true below?

File: app/tx_test.go [hunk 3]
```diff
@@ -1231,6 +1232,144 @@ func TestChangeParamspip22afterActivationHeight(t *testing.T) {
...
+	// and block size reduction
+	codec.TestMode = -4
+	// Height at which codec was upgraded from amino to proto
+	codec.UpgradeHeight = 2
+	// Height at which to enable block size upgrades
```
Comment (Olshansk) [inline app/tx_test.go]:
```suggestion
	// Height at which to enable the PerChainRTTM upgrade
```

File: app/tx_test.go [hunk 4]
```diff
@@ -1231,6 +1232,144 @@ func TestChangeParamspip22afterActivationHeight(t *testing.T) {
...
+func Test_RTTM2_ChangeParamValue(t *testing.T) {
+	cdc := memCodec()
+	ParamName := "pos/RelaysToTokensMultiplierMap"
+	ParamValue := map[string]int64{
+		"0001": 12345,
+	}
+	ExpectedValue, err := cdc.MarshalJSON(ParamValue)
+	assert.Nil(t, err)
+
+	testModeTemp := codec.TestMode
```
Comment (Olshansk) [inline app/tx_test.go:1245]:
```suggestion
	// Store temp variables for test cleanup purposes
	testModeTemp := codec.TestMode
```

File: app/tx_test.go [hunk 5]
```diff
@@ -1231,6 +1232,144 @@ func TestChangeParamspip22afterActivationHeight(t *testing.T) {
...
+	txEvent := <-txEventChan
+	assert.Equal(t, tx.TxHash, txEvent.Events["tx.hash"][0])
+	txResult, ok := txEvent.Data.(tmTypes.EventDataTx)
+	assert.True(t, ok)
+	assert.Equal(t, txResult.Result.Codespace, gov.ModuleName)
+	assert.Equal(t,
+		sdk.CodeType(txResult.Result.Code),
+		govTypes.CodeUnauthorizedParamChange,
+	)
+
```
Comment (Olshansk) [inline app/tx_test.go:1318]:
```suggestion

// Validate the param has not changed
```

File: app/tx_test.go [hunk 6]
```diff
@@ -1231,6 +1232,144 @@ func TestChangeParamspip22afterActivationHeight(t *testing.T) {
...
+	// Wait until UpgradeHeight
+	for {
+		if PCA.LastBlockHeight() >= codec.UpgradeHeight {
+			break
+		}
+		<-newBlockEventChan
+	}
+
+	height := PCA.LastBlockHeight()
```
Comment (Olshansk) [inline app/tx_test.go:1286]:
```suggestion
	// Validate the param has not changed
	height := PCA.LastBlockHeight()
```

File: codec/codec.go [hunk 7]
```diff
@@ -57,6 +57,7 @@ const (
 	VEDITKey                     = "VEDIT"
 	OutputAddressEditKey         = "OEDIT"
 	ClearUnjailedValSessionKey   = "CRVAL"
+	PerChainRTTM                 = "RTTM2"
```
Comment (Olshansk) [inline codec/codec.go]: Why do we have a "2" at the end?

File: x/gov/types/expectedKeepers.go [hunk 8]
```diff
@@ -30,6 +30,7 @@ type AuthKeeper interface {
 
 type PosKeeper interface {
 	RewardForRelays(ctx sdk.Ctx, relays sdk.BigInt, address sdk.Address)
+	RewardForRelaysPerChain(sdk.Ctx, string, sdk.BigInt, sdk.Address)
```
Comment (Olshansk) [inline x/gov/types/expectedKeepers.go]: NIT: Can you name the parameters? I'm assuming the second param is the chainID, but it's unclear.

File: x/nodes/keeper/abci.go [hunk 9]
```diff
@@ -25,7 +26,7 @@ func BeginBlocker(ctx sdk.Ctx, req abci.RequestBeginBlock, k Keeper) {
 	// store whether or not they have actually signed it and slash/unstake any
 	// which have missed too many blocks in a row (downtime slashing)
 	signedBlocksWindow := k.SignedBlocksWindow(ctx)
-	minSignedPerWindow := k.MinSignedPerWindow(ctx)
+	minSignedPerWindow := k.MinBlocksSignedPerWindow(ctx)
```
Comment (Olshansk) [inline x/nodes/keeper/abci.go:29]: In `x/nodes/keeper/params.go`, we use `MinSignedPerWindow` and here we use `MinBlocksSignedPerWindo`.

Just confirming that this is intended?

File: x/nodes/keeper/keeper.go [hunk 10]
```diff
@@ -107,8 +107,6 @@ func (k Keeper) ConvertState(ctx sdk.Ctx) {
 		panic(err)
 	}
 	k.Cdc.SetUpgradeOverride(true)
-	// custom logic for minSignedPerWindow
```
Comment (Olshansk) [inline x/nodes/keeper/keeper.go:110]: I'm not familiar with this whole manual update for MinSignedPerWindow so it's the only piece that's giving me a bit of "unease" but I don't know how to confirm/validate it.

Is your confidence level high on these changes?

File: x/nodes/keeper/reward_test.go [hunk 11]
```diff
@@ -422,3 +422,89 @@ func TestKeeper_rewardFromRelaysPIP22EXP(t *testing.T) {
 		})
 	}
 }
+
+func TestKeeper_RewardForRelaysPerChain(t *testing.T) {
+	Height_PIP22 := int64(3)
+	Height_RTTM2 := int64(10)
```
Comment (Olshansk) [inline x/nodes/keeper/reward_test.go]: DItto: What's the `2` for?

File: x/nodes/keeper/reward_test.go [hunk 12]
```diff
@@ -422,3 +422,89 @@ func TestKeeper_rewardFromRelaysPIP22EXP(t *testing.T) {
...
+	assert.True(t, rewardsDefault.Equal(rewardsNormalChain))
+
+	// Verify rewards with a non-default multiplier
+	rewardsHighProfit := keeper.RewardForRelaysPerChain(
+		ctx,
+		Chain_HighProfit,
+		NumOfRelays,
+		validator.Address,
+	)
+	assert.True(t, rewardsHighProfit.Equal(ExpectedRewards(RTTM_High)))
```
Comment (Olshansk) [inline x/nodes/keeper/reward_test.go:509]: NIT: Can you add a sanity check that `rewardsDefault` is `LT` `rewardsHighProfit`?

**Review by Olshansk — CHANGES_REQUESTED** [review]
Overall: LGTM! Great change. Simple & powerful!

1. Left some small NITs, comments and questions but nothing important. Just want to go over it before merging
2. Make sure to update the documentation (when we're closer to launch) on how to execute the tx commands with the marshalled json map
3. The ONLY concern I have is related to the cleanup you did related to the `MinSignedPerWindow`. I don't know why we had this before and wanted to understand your confidence level on it?
```
// custom logic for minSignedPerWindow
		params.MinSignedPerWindow = params.MinSignedPerWindow.QuoInt64(params.SignedBlocksWindow)
```

File: app/tx_test.go (hunk 2 above)
Comment (msmania) [reply app/tx_test.go]: It's true, but I agree it's confusing.  Will update the comment.

File: x/nodes/keeper/abci.go (hunk 9 above)
Comment (msmania) [reply x/nodes/keeper/abci.go:29]: Yes. As described in [the Part2 commit](https://github.com/pokt-network/pocket-core/pull/1580/commits/9a11177eb35da369fd0e48566dd7a888303e8558)'s message, the function `k.MinSignedPerWindow` is renamed to `k.MinBlocksSignedPerWindow`.  This function returns `int64`, the number of blocks.

The parameter value `MinSignedPerWindow` (not a function), which is `BigDec` and the current value is `0.6`, is used in only one place, `GetParams` in x/nodes/keeper/params.go.

File: app/tx_test.go (hunk 1 above)
Comment (msmania) [reply app/tx_test.go]: ok, changed these to `paramName`/`paramValue`

File: x/nodes/keeper/reward_test.go (hunk 11 above)
Comment (msmania) [reply x/nodes/keeper/reward_test.go]: I naively added 2 to distinguish per-chain RTTM from the original RTTM.  It's cryptic.  Let's change it to `PerChainRTTM`.

Comment (Olshansk) [reply x/nodes/keeper/reward_test.go]: Definitely adds a lot more sense!

File: x/nodes/keeper/keeper.go (hunk 10 above)
Comment (msmania) [reply x/nodes/keeper/keeper.go:110]: > The ONLY concern I have is related to the cleanup you did related to the MinSignedPerWindow. I don't know why we had this before and wanted to understand your confidence level on it?

For the details, please check the commit message [here](https://github.com/pokt-network/pocket-core/pull/1580/commits/9a11177eb35da369fd0e48566dd7a888303e8558).

The reason why we have this "custom logic" is the param `SignedBlocksWindow` is a proportion (= 0.6 in the mainnet), while the function `SignedBlocksWindow` is expected to return the number of blocks in a window (= 6 in the mainnet).  This custom logic converts the proportion into the number of blocks by multiplying `SignedBlocksWindow`.  This patch reconciles this discrepancy under the name `SignedBlocksWindow`.

The behavior is clear and the regression risk is low.

File: codec/codec.go (hunk 7 above)
Comment (msmania) [reply codec/codec.go]: Renamed it to `PerChainRTTM`

**Comment by Olshansk** [comment]
@msmania Ty for linking to the commit messages. I don't want this to become a "developer workflow" discussion since I know everyone has opinions/preferences on how to use git, github, IDEs, etc.

I'm sharing mine below not in expectation of others adopting it, but just so you have context/visibility into how I work and I will accommodate your workflow as well.

1. I read the GitHub PR description
2. I go through all the files 1 by 1 (see screenshot below), checking them once I'm done.
3. I use the "show changes since last review" when I do round 2, 3, etc...
4. In `GitHub` land (note: not `git`), I don't pay attention to commits at all (and don't put effort into my own commits), since we end up doing a `squash and merge` to main at the end.


Screenshot 2023-12-13 at 4 34 44 PM

Screenshot 2023-12-13 at 4 34 38 PM

**Review by Olshansk — APPROVED** [review]

**Comment by msmania** [comment]
Got it.  I personally thought one gigantic commit would be hard to be reviewed, or to be analyzed after merge in the future, so I split a patch to several PartX commits, following a convention in the projects I used to work.  Anyway, thank you for signing off!  Let's merge this!

**Comment by POKT-Discourse** [comment]
This pull request has been mentioned on **Pocket Network Forum**. There might be relevant details there:

https://forum.pokt.network/t/rc-0-11-1-upgrade-and-hi/5012/1

**Pull request merged**